src/Confluent.SchemaRegistry/Rest/RestService.cs

```diff
@@ -0,0 +1,306 @@
+    /// </remarks>
+    internal class RestService : IRestService
+    {
+        private static readonly string acceptHeader = string.Join(", ", Versions.PreferredResponseTypes);
+
+        /// <summary>
+        ///     The index of the last client successfully used (or random if none worked).
+        /// </summary>
+        private int lastClientUsed;
+        private object lastClientUsedLock = new object();
+
+        /// <summary>
+        ///     HttpClient instances corresponding to each provided schema registry Uri.
+        /// </summary>
+        private readonly List<HttpClient> clients;
+
+        
+        /// <summary>
+        ///     Initializes a new instance of the RestService class.
+        /// </summary>
+        public RestService(string schemaRegistryUrl, int timeoutMs)
+        { 
+            this.clients = schemaRegistryUrl
+                .Split(',')
+                .Select(uri => uri.StartsWith("http", StringComparison.Ordinal) ? uri : "http://" + uri) // need http or https - use http if not present.
+                .Select(uri => new HttpClient() { BaseAddress = new Uri(uri, UriKind.Absolute), Timeout = TimeSpan.FromMilliseconds(timeoutMs) })
+                .ToList();
+        }
+
+        #region Base Requests
+
+        private async Task<HttpResponseMessage> ExecuteOnOneInstanceAsync(HttpRequestMessage request)
+        {
+            // There may be many base urls - roll until one is found that works.
+            //
+            // Start with the last client that was used by this method, which only gets set on 
+            // success, so it's probably going to work.
+            //
+            // Otherwise, try every client until a successful call is made (true even under 
+            // concurrent access).
+
+            string aggregatedErrorMessage = null;
+            HttpResponseMessage response = null;
+            bool gotOkAnswer = false;
+            bool firstError = true;
+            
+            int startClientIndex;
+            lock (lastClientUsedLock)
+            {
+                startClientIndex = this.lastClientUsed;
+            }
+
+            int loopIndex = startClientIndex;
+            int clientIndex = -1; // prevent uninitialized variable compiler error.
+            for (; loopIndex < clients.Count + startClientIndex && !gotOkAnswer; ++loopIndex)
+            {
+                clientIndex = loopIndex % clients.Count;
+
+                try
+                {
+                    response = await clients[clientIndex]
+                            .SendAsync(request).ConfigureAwait(false);
+
+                    // In the case of an internal server error, try another server 
+                    //   (reason could be e.g. "error while forwarding the request to the master")
+                    // 4xx errors should not be retried (these are conclusive)
+                    if (response.StatusCode == HttpStatusCode.InternalServerError)
```

Comment:
check for 500 range rather than 500 specifiicificialciacally

Reply:
yes, I think that makes sense.